### PR TITLE
feat: Add Regex matching support to IOx's gRPC API

### PR DESCRIPTION
**Pull request by e-dard**
Closes https://github.com/influxdata/conductor/issues/281

This PR adds support for regex matching via the gRPC API. Effectively, this teaches IOx (and Data Fusion really) how to filter results via a predicate expression that is a match (or not a match) to a regular expression.

In Flux we would express this as:

```
from(bucket: "edd")
  |> range(....)
  |> filter(fn: (r) => r["_measurement"] == "cpu")
  ...
  |> filter(fn: (r) => r["city"] =~ /^new.*/)
```

This PR adds support for both regex matching (`=~`) and also negated matching (`!~`)`. Negated matching effectively means "filter away any rows with values that match the regular expression".

I have tested this PR locally using IOx and Flux. 

Kapture 2021-05-06 at 19 32 12 2

This PR may be easier to review commit-by-commit:

- bf8d70b: adds a new user defined function to DataFusion that allows us to apply a regex to every value in an Arrow string array, producing a `BooleanArray` that signifies if each values in the array satisfied the regular expression.
- 8436742: teaches the InfluxRPC front-end how to build the UDF into a regular `query::Predicate` as a filter expression.
- 516840b: teaches the gRPC API server how to convert regular expression operations sent from Flux/InfluxQL into our custom DataFusion ScalarFunction (the UDF). 
-

### CODE REVIEW

File: query/src/func/regex.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+///
+/// If `matches` is true then this expression will filter values that do not
+/// satisfy the regex (equivalent to `col ~= /pattern/`). If `matches` is `false`
+/// then the expression will filter values that *do* match the regex, which is
+/// equivalent to `col !~ /pattern/`.
+///
+/// This UDF is designed to support the regex operator that can be pushed down
+/// via the InfluxRPC API.
+///
+pub fn regex_match_expr(input: Expr, pattern: String, matches: bool) -> Expr {
```
Comment (e-dard) [inline query/src/func/regex.rs:29]: This function creates a DataFusion `ScalarUDF`, which basically is a function that can be handed Arrow `StringArray`s and will apply a regular expression to them to produce a `BooleanArray` of matches. 

You might be asking why not use the Arrow regex kernel? Well that returns the matches themselves, which isn't really what DataFusion wants for the purposes of filtering.

File: query/src/func/regex.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    let udf_name = if matches {
+        REGEX_MATCH_UDF_NAME
+    } else {
+        REGEX_NOT_MATCH_UDF_NAME
+    };
+
+    let udf = create_udf(
+        udf_name,
+        vec![DataType::Utf8],
+        // Arc::new(DataType::Utf8),
```
Comment (e-dard) [inline query/src/func/regex.rs:70]:
```suggestion
        // Arc::new(DataType::Utf8),
```

Cruft. Will remove

File: query/src/func/regex.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                    "| aphex twin    | 10     |",
+                    "| bruce         | 5      |",
+                    "| Blood Orange  | 12     |",
+                    "| cocteau twins | 13     |",
+                    "+---------------+--------+",
+                    "",
+                ],
+            ),
+            (
+                ".+O.*", // match just words containing "O".
```
Comment (e-dard) [inline query/src/func/regex.rs:136]: We can filter on all sorts of fanciness! 🎉

File: server/src/query_tests/influxrpc/read_filter.rs [hunk 4]
```diff
@@ -272,6 +272,80 @@ async fn test_read_filter_data_pred_refers_to_good_and_non_existent_columns() {
     run_read_filter_test_case!(TwoMeasurements {}, predicate, expected_results);
 }
 
+#[tokio::test]
+async fn test_read_filter_data_pred_using_regex_match() {
```
Comment (e-dard) [inline server/src/query_tests/influxrpc/read_filter.rs:276]: Here we are exercising using the new regex UDF for filtering via the InfluxRPC frontend.

File: src/influxdb_ioxd/rpc/storage/expr.rs [hunk 5]
```diff
@@ -471,16 +468,16 @@ fn build_logical_node(logical: i32, inputs: Vec<Expr>) -> Result<Expr> {
...
 fn build_comparison_node(comparison: i32, inputs: Vec<Expr>) -> Result<Expr> {
     let comparison_enum = RPCComparison::from_i32(comparison);
 
     match comparison_enum {
         Some(RPCComparison::Equal) => build_binary_expr(Operator::Eq, inputs),
         Some(RPCComparison::NotEqual) => build_binary_expr(Operator::NotEq, inputs),
         Some(RPCComparison::StartsWith) => StartsWithNotSupported {}.fail(),
-        Some(RPCComparison::Regex) => RegExpNotSupported {}.fail(),
-        Some(RPCComparison::NotRegex) => NotRegExpNotSupported {}.fail(),
+        Some(RPCComparison::Regex) => build_regex_match_expr(true, inputs),
```
Comment (e-dard) [inline src/influxdb_ioxd/rpc/storage/expr.rs:479]: Now we are right out at the gRPC API layer. We need to take deserialised protobuf objects and conver them into our internal DataFusion UDF objects.

File: src/influxdb_ioxd/rpc/storage/expr.rs [hunk 6]
```diff
@@ -505,6 +502,24 @@ fn build_binary_expr(op: Operator, inputs: Vec<Expr>) -> Result<Expr> {
...
+    let num_children = inputs.len();
+    match num_children {
+        2 => {
+            let pattern = if let Expr::Literal(ScalarValue::Utf8(pattern)) = inputs.remove(1) {
+                pattern.context(RegExpPatternInvalid)?
+            } else {
+                return InternalInvalidRegexExprReference.fail();
+            };
+
+            Ok(regex::regex_match_expr(inputs.remove(0), pattern, matches))
```
Comment (e-dard) [inline src/influxdb_ioxd/rpc/storage/expr.rs:517]: I wasn't very keen on reaching into the `query::func::regex` module from here but didn't see an alternative that was much better.

File: tests/end_to_end_cases/storage_api.rs [hunk 7]
```diff
@@ -294,6 +294,47 @@ async fn measurement_fields_endpoint(
     assert_eq!(field.timestamp, scenario.ns_since_epoch() + 4);
 }
 
+#[tokio::test]
+pub async fn regex_operator_test() {
```
Comment (e-dard) [inline tests/end_to_end_cases/storage_api.rs:298]: Exercising the gRPC API with an end to end test.

File: tests/end_to_end_cases/storage_api.rs [hunk 8]
```diff
@@ -294,6 +294,47 @@ async fn measurement_fields_endpoint(
...
+    let read_source = scenario.read_source();
+
+    // read_group(group_keys: region, agg: None)
+    let read_filter_request = ReadFilterRequest {
+        read_source: read_source.clone(),
+        range: Some(TimestampRange {
+            start: 0,
+            end: 2001, // include all data
+        }),
+        predicate: Some(make_regex_match_predicate("host", "^b.+")),
```
Comment (e-dard) [inline tests/end_to_end_cases/storage_api.rs:318]: Let's get all frames where `host =~ /^b.+/`.

**Review by alamb — APPROVED** [review]
I think it looks really nice @e-dard . 👍  I don't have anything of substance to add

I went over it fairly carefully and it all looks very nice

File: query/src/func/regex.rs [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+        error::DataFusionError,
+        logical_plan::{col, Expr},
+        prelude::ExecutionContext,
+    };
+    use std::iter::FromIterator;
+    use std::sync::Arc;
+
+    #[tokio::test]
+    async fn regex_match_expr() {
+        let cases = vec![
```
[the remark this answers is not in the document]
Comment (e-dard) [reply query/src/func/regex.rs:97]: I was going to have more complex predicates on both columns but then didn't see the need.

File: query/src/func/regex.rs [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+};
+use datafusion::{
+    error::DataFusionError,
+    logical_plan::{create_udf, Expr},
+    physical_plan::functions::make_scalar_function,
+};
+
+/// The name of the regex_match UDF given to DataFusion.
+pub const REGEX_MATCH_UDF_NAME: &str = "RegexMatch";
+pub const REGEX_NOT_MATCH_UDF_NAME: &str = "RegexMatch";
```
Comment (domodwyer) [inline query/src/func/regex.rs:15]: Are these being the same a copy/paste mishap? It makes some of the if's (such as in the predicate support visitor and UDF) a bit redundant.

File: query/src/func/regex.rs [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+                    "+---------------+--------+",
+                    "| Blood Orange  | 12     |",
+                    "| cocteau twins | 13     |",
+                    "+---------------+--------+",
+                    "",
+                ],
+            ),
+        ];
+
+        for (pattern, matches, expected) in cases.into_iter() {
```
Comment (domodwyer) [inline query/src/func/regex.rs:162]: Ah table tests - how I miss you and your clean, satisfying setup! Nice 👍 

I've started using macros more and more to generate test funcs given a test case as input each time - it's slightly less lovely than the pure test case array you have here, but it has the upside of if one test case fails on an unwrap or something, it has a distinct test name so you know which test case failed. I really wish there was a `t.Run()` equivalent!

**Review by domodwyer — APPROVED** [review]
I really enjoyed reading this PR - it highlighted how the internal bits are stitched together for me!

File: query/Cargo.toml [hunk 12]
```diff
@@ -27,7 +27,8 @@ influxdb_line_protocol = { path = "../influxdb_line_protocol" }
 internal_types = { path = "../internal_types" }
 metrics = { path = "../metrics" }
 parking_lot = "0.11.1"
-snafu = "0.6.2"
+regex = "1"
+snafu = "0.4.3"
```
[the remark this answers is not in the document]
Comment (e-dard) [reply query/Cargo.toml:31]: I did not. Weird

File: query/src/func/regex.rs (hunk 11 above)
Comment (alamb) [reply query/src/func/regex.rs:161]: I use a similar approach in the `query_test`s -- make a macro that does the work (and also generates errors on the right line) https://github.com/influxdata/influxdb_iox/blob/main/server/src/query_tests/sql.rs#L16

**Pull request merged**